src/main/java/net/sf/jabref/logic/util/io/FileUtil.java

```diff
@@ -318,4 +321,125 @@ private static File shortenFileName(File fileName, String dir) {
             return fileName;
         }
     }
+
+    public static Map<BibEntry, List<File>> findAssociatedFiles(Collection<BibEntry> entries, Collection<String> extensions, Collection<File> directories) {
+        HashMap<BibEntry, List<File>> result = new HashMap<>();
+
+        // First scan directories
+        Set<File> filesWithExtension = FileFinder.findFiles(extensions, directories);
+
+        // Initialize Result-Set
+        for (BibEntry entry : entries) {
+            result.put(entry, new ArrayList<>());
+        }
+
+        boolean exactOnly = Globals.prefs.getBoolean(JabRefPreferences.AUTOLINK_EXACT_KEY_ONLY);
+        // Now look for keys
+        nextFile: for (File file : filesWithExtension) {
+
+            String name = file.getName();
+            int dot = name.lastIndexOf('.');
+            // First, look for exact matches:
+            for (BibEntry entry : entries) {
+                String citeKey = entry.getCiteKey();
+                if ((citeKey != null) && !citeKey.isEmpty()) {
+                    if (dot > 0) {
+                        if (name.substring(0, dot).equals(citeKey)) {
+                            result.get(entry).add(file);
+                            continue nextFile;
+                        }
+                    }
+                }
+            }
+            // If we get here, we didn't find any exact matches. If non-exact
+            // matches are allowed, try to find one:
+            if (!exactOnly) {
+                for (BibEntry entry : entries) {
+                    String citeKey = entry.getCiteKey();
+                    if ((citeKey != null) && !citeKey.isEmpty()) {
+                        if (name.startsWith(citeKey)) {
+                            result.get(entry).add(file);
+                            continue nextFile;
+                        }
+                    }
+                }
+            }
+        }
+
+        return result;
+    }
+
+    public static List<List<String>> decodeFileField(String value) {
```

Comment:
It can't be deleted (but you are otherwise correct), as the added code is much more limited. However, with some thinking the code in FileListTableModel can be simplified and the added code used (primary problem is the code "optimized" for getting the first item only).

The code in FileListTableModel uses FileListEntry which consists of description, link, and ExternalFileType. Although ExternalFileType resides in external, so FileListEntry only includes from there, ExternalFileType has both a JPanel and an Icon. Although, it was my original plan to extract the whole thing setting up a list of FileListEntry, I thought that introducing (hidden) dependencies on Swing/GUI components in logic was a bad idea (although the independence tests might have passed).

One could possibly introduce another class, FileItem(?) which consists of the description, link, and filetype name. This is really enough for serialization and may be enough in some of the cases where now the whole FileListTableModel machinery is included (as you state in #572).